Add error_on_line_overflow_strings config option

topecongiro opened the pull request:
This PR adds a new config option `error_on_line_overflow_strings`, which is a variant of `error_on_line_overflow_comments`.

EDIT: closes #2268.

topecongiro commented:
I have a few questions:

1. `error_on_line_overflow_strings` and `error_on_line_overflow_strings` suppress warnings on trailing white spaces as well. To make things clear, should we change the name of two options?
2. Should we combine the two options?
3. Should we turn them off (set to `false`) by default?

CAD97 commented on src/config.rs:
```diff
@@ -674,6 +674,8 @@ create_config! {
     error_on_line_overflow_comments: bool, true, false,
         "Error if unable to get comments within max_width";
+    error_on_line_overflow_strings: bool, true, false,
+        "Error if unable to get string letrais within max_width";
```
`letrais` does not match any dictionary word.

Did you mean one of:

- `letters`
- `literal`
- `laterals`
- `literals`
- `lateral`

?

(Sorry, just having fun. Those are legitimate suggestions from a spell checker, though. You meant `literals` here.)

topecongiro replied:
Thank you, this will make my typo of the year for 2017.

nrc commented:
>    error_on_line_overflow_strings and error_on_line_overflow_strings suppress warnings on trailing white spaces as well. To make things clear, should we change the name of two options?

Yeah, I think `error_on_unformatted` would be good, to cover everything.

>   Should we combine the two options?

Yes, I can't imagine someone wanting one, but not the other

>   Should we turn them off (set to false) by default?

Hmm, not sure about the default. I think probably yes the option should be off by default. But we should be able to turn it on via the command line (for the CI use case).

topecongiro commented:
Updated the PR with the review reflected and rebased against master.

1. Combined the two options into `error_on_unformatted_comments_or_strings` (too long?)
2. Added `--error-on-unformatted` command line option.
3. Sorted command line options in alphabetical order.

```


usage: ./target/debug/rustfmt [options] <file>...

Options:
        --color [always|never|auto]
                        Use colored output (if supported)
        --config-help   Show details of rustfmt configuration options
        --config-path [Path for the configuration file]
                        Recursively searches the given path for the
                        rustfmt.toml config file. If not found reverts to the
                        input file path
        --dump-default-config PATH
                        Dumps default configuration to PATH. PATH defaults to
                        stdout, if omitted.
        --dump-minimal-config PATH
                        Dumps configuration options that were checked during
                        formatting to a file.
        --error-on-unformatted 
                        Error if unable to get comments or string literals
                        within max_width, or they are left with trailing
                        whitespaces
        --file-lines JSON
                        Format specified line ranges. See README for more
                        detail on the JSON format.
    -h, --help          Show this message
        --skip-children 
                        Don't reformat child modules
        --unstable-features 
                        Enables unstable features. Only available on nightly
                        channel
    -v, --verbose       Print verbose output
    -V, --version       Show version information
        --write-mode [replace|overwrite|display|plain|diff|coverage|checkstyle]
                        How to write output (not usable when piping from
                        stdin)

Exit Codes:
    0 = No errors
    1 = Encountered operational errors e.g. an IO error
    2 = Failed to reformat code because of parsing errors
    3 = Code is valid, but it is impossible to format it properly
    4 = Formatted code differs from existing code (write-mode diff only)
```

nrc commented:
> Combined the two options into error_on_unformatted_comments_or_strings (too long?)

Could we use just `error_on_unformatted` for the option too? I think it would be good to be consistent between the option and command line flag

nrc commented:
r+ with that last change

topecongiro commented:
Thanks again, rename the option.

topecongiro commented:
Rebased against master.

nrc commented:
Thanks!

For the future, it would be great if you could squash some of this kind of commit - exactly how to structure commits in a PR is more art than science (and yours are usually very nice, and I am not perfect by a long way). I would squash typo and minor refactoring commits into commits making larger changes, though I would try and keep refactoring commits separate from those that change behaviour.  The best guide, I think, is to consider if when reading blame for the repo, whether it is useful to have the code in a separate commit - to take some examples from this PR - 'Print command line options in alphabetical order' and 'Add a config option to suppress error message on string literal' - definitely yes, 'Fix a typo ' - definitely no, 'Organize command line options and start with upper case' - maybe or maybe better squashed with 'Print command line options in alphabetical order', depending on the code. Anyway, this doesn't have to be perfect, and I wouldn't spend too much time on it, but there might be a few places you can make a better commit series without too much stress.

topecongiro commented:
Thank you the additional advice, this is very helpful!
